build/parseList.c

```diff
@@ -8,13 +8,86 @@
 #include "build/rpmbuild_internal.h"
 #include "debug.h"
 
+static int addLinesFromFile(rpmSpec spec, const char * const fn, rpmTagVal tag)
+{
+    int nlines = 0;
+    int flags = STRIP_COMMENTS | STRIP_TRAILINGSPACE | ALLOW_EMPTY;
+    ARGV_t argv = NULL;
+
+    nlines = readManifest(spec, fn, tag == RPMTAG_SOURCE ? "%sourcelist" : "%patchlist", flags, &argv, NULL);
+
+    for (ARGV_t av = argv; *av; av++) {
+	addSource(spec, 0, *av, tag);
+    }
+
+    argvFree(argv);
+
+    if (nlines == 0) {
+	rpmlog(RPMLOG_WARNING, _("Empty %s file %s\n"),
+	       tag == RPMTAG_SOURCE ? "%sourcelist" : "%patchlist",
+	       fn);
+    }
+
+    return (nlines < 0) ? RPMRC_FAIL : RPMRC_OK;
+}
 
 int parseList(rpmSpec spec, const char *name, rpmTagVal tag)
 {
     int res = PART_ERROR;
     ARGV_t lst = NULL;
+    int rc, argc = 0;
+    int arg;
+    const char **argv = NULL;
+    poptContext optCon = NULL;
+    struct poptOption optionsTable[] = {
+	{ NULL, 'f', POPT_ARG_STRING, NULL, 'f', NULL, NULL},
+	{ 0, 0, 0, 0, 0, NULL, NULL}
+    };
+    char * file = NULL;
+
+    if ((rc = poptParseArgvString(spec->line, &argc, &argv))) {
+	rpmlog(RPMLOG_ERR, _("line %d: Error parsing %%%s: %s\n"),
+	       spec->lineNum, name, poptStrerror(rc));
+	goto exit;
+    }
+
+    optCon = poptGetContext(NULL, argc, argv, optionsTable, 0);
+    while ((arg = poptGetNextOpt(optCon)) > 0) {
+
+	char * filename = poptGetOptArg(optCon);
+	if (!filename) {
+	    rpmlog(RPMLOG_ERR,
+		   _("line %d: \"%%%s -f\" requires an argument.\n"),
+		   spec->lineNum, name);
+	    goto exit;
+	}
+
+	addSource(spec, 0, filename, RPMTAG_SOURCE);
```

Comment:
Mm. This is out of line with our -f usage elsewhere: we don't ask questions about the file origin external sources are pulled in, whether that's %include'd or pulled via -f from in scriptlet, files manifest or otherwise. Much less make them our own.

I can see a use-case for at least %patchlist -f (not so sure about %sourcelist -f), but these external entities are problematic in general as they break the standalone expectations that people have on specs. %include is notoriously nasty, so the question is do we really want more of that?

Reply:
So what are you suggesting here beyond dropping the addSource()? Not supporting -f for %sourcelist?

Thinking a bot more about this: The files containing the patches and sources need to be there right at the beginning of the build - before %prep. So there are not that many opportunities to create them during build. So one wonders what to do with them other than adding them to the sources. We might need to check that they are in the right place.
But if they are not part of the sources installing and building the SRPM will surely fail.

Reply:
With all the other -f's it's intentionally left to be somebody elses headache: if you want it to be shipped with your srpm, you manually add a Source: for it, but if not then its assumed to be something from the build environment, one way or the other. Of course most of the time such things should be specified as sources. So adding a source for the file in which your sources are defined ... this gets a bit crazy.

And yeah, unlike %files -f, these cannot really be created by the build itself. Which means this is for most practical purposes just a specialized version of %include, with many/most of the same problems.

So what I'm really saying here is that we need to think this whole thing through properly before digging further into implementation details.